Gateway/Request/AdditionalDataLevel23DataBuilder.php

```diff
@@ -0,0 +1,118 @@
+<?php
+/**
+ *                       ######
+ *                       ######
+ * ############    ####( ######  #####. ######  ############   ############
+ * #############  #####( ######  #####. ######  #############  #############
+ *        ######  #####( ######  #####. ######  #####  ######  #####  ######
+ * ###### ######  #####( ######  #####. ######  #####  #####   #####  ######
+ * ###### ######  #####( ######  #####. ######  #####          #####  ######
+ * #############  #############  #############  #############  #####  ######
+ *  ############   ############  #############   ############  #####  ######
+ *                                      ######
+ *                               #############
+ *                               ############
+ *
+ * Adyen Payment Module
+ *
+ * Copyright (c) 2022 Adyen N.V.
+ * This file is open source and available under the MIT license.
+ * See the LICENSE file for more info.
+ *
+ * Author: Adyen <magento@adyen.com>
+ */
+
+namespace Adyen\Payment\Gateway\Request;
+
+use Adyen\Payment\Helper\ChargedCurrency;
+use Adyen\Payment\Helper\Config;
+use Adyen\Payment\Helper\Data;
+use Adyen\Util\Uuid;
+use Magento\Payment\Gateway\Helper\SubjectReader;
+use Magento\Payment\Gateway\Request\BuilderInterface;
+use Magento\Sales\Model\Order;
+use Magento\Sales\Model\Order\Item;
+use Magento\Store\Model\StoreManagerInterface;
+
+class AdditionalDataLevel23DataBuilder implements BuilderInterface
+{
+    /**
+     * @var Config
+     */
+    private $config;
+    /**
+     * @var StoreManagerInterface
+     */
+    private $storeManager;
+    /**
+     * @var Data
+     */
+    private $adyenHelper;
+    /**
+     * @var ChargedCurrency
+     */
+    private $chargedCurrency;
+
+    public function __construct(
+        Config $config,
+        StoreManagerInterface $storeManager,
+        ChargedCurrency $chargedCurrency,
+        Data $adyenHelper
+    )
+    {
+        $this->config = $config;
+        $this->storeManager = $storeManager;
+        $this->adyenHelper = $adyenHelper;
+        $this->chargedCurrency = $chargedCurrency;
+    }
+
+    public function build(array $buildSubject)
+    {
+        $requestBody = [];
+        if ($this->config->sendLevel23AdditionalData($this->storeManager->getStore()->getId())) {
+            $paymentDataObject = SubjectReader::readPayment($buildSubject);
+            $payment = $paymentDataObject->getPayment();
+            /** @var Order $order */
+            $order = $payment->getOrder();
+            $currencyCode = $this->chargedCurrency->getOrderAmountCurrency($order)->getCurrencyCode();
+
+            if (!$order->getCustomerIsGuest()) {
+                $customerReference = str_pad($order->getCustomerId(), 3, '0', STR_PAD_LEFT);
+            } else {
+                $uuid = Uuid::generateV4();
```

Comment:
For guest users I think this also gets generated in another builder, should we generate it in a single place so that there's no mismatch?

Reply:
27afb05